File: maze-walker/typescript/src/Walker.ts
import { type Cell, cellKey, parseKey, cardinalNeighbors } from './Cell.js';
import { type Maze } from './Maze.js';

/**
 * Navigates a maze using BFS to find the shortest path from start to end.
 */
export class Walker {
  private readonly maze: Maze;

  constructor(maze: Maze) {
    this.maze = maze;
  }

  /**
   * Finds the shortest path from start to end using BFS.
   * Returns an empty array when no path exists.
   */
  findPath(): Cell[] {
    const { start, end } = this.maze;
    const startKey = cellKey(start);
    const endKey = cellKey(end);

    const visited = new Set<string>([startKey]);
    const queue: string[] = [startKey];
    const cameFrom = new Map<string, string>();

    let head = 0;
    while (head < queue.length) {
      const currentKey = queue[head]!;
      head++;

      if (currentKey === endKey) {
        return this.reconstructPath(cameFrom, startKey, endKey);
      }

      const current = parseKey(currentKey);
      for (const neighbor of cardinalNeighbors(current)) {
        if (!this.maze.isWalkable(neighbor)) continue;

        const nk = cellKey(neighbor);
        if (visited.has(nk)) continue;

        visited.add(nk);
        cameFrom.set(nk, currentKey);
        queue.push(nk);
      }
    }

    return [];
  }

  private reconstructPath(
    cameFrom: Map<string, string>,
    startKey: string,
    endKey: string,
  ): Cell[] {
    const path: Cell[] = [];
    let current = endKey;

    while (current !== startKey) {
      path.push(parseKey(current));
      current = cameFrom.get(current)!;
    }

    path.push(parseKey(startKey));
    path.reverse();
    return path;
  }
}

File: maze-walker/python/src/maze_walker/walker.py
from __future__ import annotations

from collections import deque

from .cell import Cell
from .maze import Maze


class Walker:
    """Navigates a maze using BFS to find the shortest path from start to end."""

    def __init__(self, maze: Maze) -> None:
        self._maze = maze

    def find_path(self) -> list[Cell]:
        """
        Finds the shortest path from start to end using BFS.
        Returns an empty list when no path exists.
        """
        start = self._maze.start
        end = self._maze.end

        visited: set[Cell] = {start}
        queue: deque[Cell] = deque([start])
        came_from: dict[Cell, Cell] = {}

        while queue:
            current = queue.popleft()

            if current == end:
                return self._reconstruct_path(came_from, start, end)

            for neighbor in current.cardinal_neighbors():
                if not self._maze.is_walkable(neighbor):
                    continue
                if neighbor in visited:
                    continue

                visited.add(neighbor)
                came_from[neighbor] = current
                queue.append(neighbor)

        return []

    @staticmethod
    def _reconstruct_path(
        came_from: dict[Cell, Cell], start: Cell, end: Cell
    ) -> list[Cell]:
        path: list[Cell] = []
        current = end

        while current != start:
            path.append(current)
            current = came_from[current]

        path.append(start)
        path.reverse()
        return path
